Desktop_Interface/functiongencontrol.cpp

```diff
@@ -5,126 +5,56 @@
-
-    directory.append("/waveforms/");
-    directory.append(newName);
-    QByteArray temp = directory.toLocal8Bit();
-    char *fileName = temp.data();
-
-    qDebug() << "opening" << fileName;
-    FILE *fptr = fopen(fileName, "r");
-        if (fptr==NULL){
-            qFatal("%s could not be opened!", fileName);
-        }
-
-    char lengthString[6];
-    fgets(lengthString, 5, fptr);
-    qDebug() << lengthString;
-    sscanf(lengthString, "%d", &length_CH1);
-
-    char divisibilityString[6];
-    //Bit of bullshit to deal with CRLF line endings on Mac.
-    do fgets(divisibilityString, 5, fptr);
-    while ((divisibilityString[0] == '\r') || (divisibilityString[0] == '\n'));
-    sscanf(divisibilityString, "%d", &divisibility_CH1);
-
-    qDebug() << "Length = " << length_CH1;
-    qDebug() << "Divisibility = " << divisibility_CH1;
-
-    free(samples_CH1);
-    samples_CH1 = (unsigned char *) malloc(length_CH1);
-
-    char *dataString = (char *) malloc(length_CH1*5+1);
-    fgets(dataString, length_CH1*5+1, fptr);
-
-    int dummy;
-    char *dataStringCurrent = dataString;
-    for (int i=0;i<length_CH1;i++){
-        sscanf(dataStringCurrent, "%d", &dummy);
-        dataStringCurrent += strcspn(dataStringCurrent, "\t") + 1;
-        samples_CH1[i] = (unsigned char) dummy;
-    }
-
-    free(dataString);
-    fclose(fptr);
-#endif
-    setMaxFreq_CH1(DAC_SPS/(length_CH1>>(divisibility_CH1-1)));
-    setMinFreq_CH1((double) CLOCK_FREQ/1024/65535/length_CH1);
-    functionGenToUpdate(0, this);
+	waveformName<6>(newName, CH1, 0, [](QString const& str){return str.toLocal8Bit(); });
 }
 
 void functionGenControl::freqUpdate_CH1(double newFreq){
     qDebug() << "newFreq = " << newFreq;
-    freq_CH1 = newFreq;
+	CH1.freq = newFreq;
     functionGenToUpdate(0, this);
 }
 
 void functionGenControl::amplitudeUpdate_CH1(double newAmplitude){
     qDebug() << "newAmplitude = " << newAmplitude;
-    amplitude_CH1 = newAmplitude;
+	CH1.amplitude = newAmplitude;
     functionGenToUpdate(0, this);
 }
 
 void functionGenControl::offsetUpdate_CH1(double newOffset){
     qDebug() << "newOffset = " << newOffset;
-    offset_CH1 = newOffset;
+	CH1.offset = newOffset;
     functionGenToUpdate(0, this);
 }
 
 void functionGenControl::waveformName_CH2(QString newName)
+{
+	waveformName<5>(newName, CH2, 1, [](QString const& str){ return str.toLatin1(); });
+}
+
+void functionGenControl::freqUpdate_CH2(double newFreq){
+    qDebug() << "newFreq2 = " << newFreq;
+	CH2.freq = newFreq;
+    functionGenToUpdate(1, this);
+}
+
+void functionGenControl::amplitudeUpdate_CH2(double newAmplitude){
+    qDebug() << "newAmplitude2 = " << newAmplitude;
+	CH2.amplitude = newAmplitude;
+    functionGenToUpdate(1, this);
+}
+
+void functionGenControl::offsetUpdate_CH2(double newOffset){
+    qDebug() << "newOffset2 = " << newOffset;
+	CH2.offset = newOffset;
+    functionGenToUpdate(1, this);
+}
+
+template<int LengthStringSize, typename Lambda>
```

Comment:
LengthString and DivisibilityString are just small buffers that hold strings to be read in from a file.  The strings should be 4 characters long or less, hence the size of 5.  That being said, you can easily make the length higher (if you look at the Android stuff, it uses 16-byte buffers) and it won't cause any issues.  After all, stack size is something like 8MiB, what difference does 11 bytes make?

Reply:
I was thinking just that. That change is in!